# Challenge 2

In [1]:
import torch
if torch.cuda.is_available():
  print("GPU used")
else:
  print("CPU used, slower execution")
  print("Try to get a GPU runtime for faster progress")

GPU used


In [2]:
!pwd
!ls

Der Befehl "pwd" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.
Der Befehl "ls" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.


In [3]:
remove_project = False
if remove_project:
  # go to default directory - for Colab /content
  %cd /content

  !rm -rf Challenge2
!pwd
!ls

Der Befehl "pwd" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.
Der Befehl "ls" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.


In [4]:
!git clone https://github.com/Ma0256/Challenge2.git

Cloning into 'Challenge2'...


In [5]:
%cd Challenge2
!ls

C:\Users\ma0256\Sandbox\Challenge2


Der Befehl "ls" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.


In [7]:
# installing all reqs can take a long time. First try to run with default package installation.
#!pip install -r requirements.txt

## Download Dataset

In [8]:
from dataset.dataset_ESC50 import ESC50
import config

testset = ESC50(subset="test", root=config.esc50_path, download=True)

data/esc50/master.zip: 616MiB [00:40, 16.1MiB/s]


This dataset contains 2000 audio files and is pre-organized in 5 equal sized folds of 400 files. The predictive performance is measured via average accuracy of 5-fold cross validation.

What is the expected accuracy for random guessing? Check the number of classes.

In [10]:
config.n_classes

50

## Evaluate Cross Validation
This will test all 5 cross validation folds with training results of a sample-run - every fold with the respective model. First the sample-run (incl. the models) is downloaded. Then the model is loaded with the sample-run parameters and tested on the fold, for all 5 folds.

In [11]:
import pandas as pd
!python test_crossval.py

download model params
results/sample-run.zip: 100% 197M/197M [00:23<00:00, 8.82MiB/s]
*****
ResNet(block=ResidualBlock, layers=[2, 2, 2, 2], num_classes=config.n_classes)
*****

test results/sample-run/1/terminal.pt
100% 13/13 [00:58<00:00,  4.52s/bat]
TestAcc    0.685
TestLoss   1.155
dtype: float64

test results/sample-run/2/terminal.pt
100% 13/13 [00:45<00:00,  3.54s/bat]
TestAcc    0.665
TestLoss   1.173
dtype: float64

test results/sample-run/3/terminal.pt
100% 13/13 [00:46<00:00,  3.57s/bat]
TestAcc    0.688
TestLoss   1.166
dtype: float64

test results/sample-run/4/terminal.pt
100% 13/13 [00:45<00:00,  3.51s/bat]
TestAcc    0.790
TestLoss   0.814
dtype: float64

test results/sample-run/5/terminal.pt
100% 13/13 [00:47<00:00,  3.65s/bat]
TestAcc    0.688
TestLoss   1.169
dtype: float64
     terminal.pt         
         TestAcc TestLoss
1          0.685    1.155
2          0.665    1.173
3          0.688    1.166
4          0.790    0.814
5          0.688    1.169
mean       0.703

The mean CV accuracy should be TestAcc=0.703. This is quite good, given a random baseline of 0.02 / 2%.

The sample-run directory now contains two CSV files, one with test scores and another with predictions.

In [31]:
!ls results/sample-run/

1  2  3  4  5  test_probs_terminal.csv	test_scores_terminal.csv


Let's check the scores.

In [47]:
pd.read_csv('results/sample-run/test_scores_terminal.csv', index_col=0)

TestAcc  TestLoss
1     0.685000  1.154911
2     0.665000  1.172668
3     0.687500  1.166192
4     0.790000  0.814113
5     0.687500  1.168974
mean  0.703000  1.095371
std   0.049542  0.157368

The predictions file is for submission. For every tested WAV file it contains the model predictions for the 50 classes.

In [49]:
pd.read_csv('results/sample-run/test_probs_terminal.csv', index_col=0).head()

0         1         2         3         4         5  \
1-100032-A-0.wav   4.425216  2.618636  1.091619  0.937591 -1.214858  0.475062   
1-100038-A-14.wav -1.364774 -1.451071  0.235597  1.069462  1.050353 -1.875610   
1-100210-A-36.wav -2.595013  0.955836  0.729774 -0.285057 -0.706471 -0.265884   
1-100210-B-36.wav -3.005787  1.575030  0.815861 -0.552081 -1.507387  0.194218   
1-101296-A-19.wav -1.446895  0.196831 -1.059902  2.147269 -2.860927  0.752272   

                          6         7         8         9  ...        40  \
1-100032-A-0.wav  -1.795039 -1.223864 -0.957428  2.538039  ...  1.118637   
1-100038-A-14.wav -3.121527 -0.687590  1.195842  3.208323  ...  1.302160   
1-100210-A-36.wav -2.161351 -1.215864  1.662850  4.394828  ...  1.259001   
1-100210-B-36.wav -2.077039 -0.658207  1.641086  2.470541  ...  0.705501   
1-101296-A-19.wav -3.299027  1.993068  2.246752 -0.155412  ...  3.810448   

                         41        42        43        44        45        46  \
1-100032-A-0.wav  -1.829709 -1.933142  0.755511 -0.964274 -1.923269 -2.302734   
1-100038-A-14.wav  2.325430 -1.037800  3.077839 -1.053247  0.803705 -0.291807   
1-100210-A-36.wav  3.635612  0.831326  2.440547  0.870062 -0.725592 -0.766259   
1-100210-B-36.wav  3.461071  1.730435  2.282524  0.890644 -0.626794 -0.259676   
1-101296-A-19.wav  0.690591  0.213421  0.255530 -2.173933  4.546542 -0.306779   

                         47        48        49  
1-100032-A-0.wav  -1.352861 -0.435046 -1.488122  
1-100038-A-14.wav  0.505203  0.014139  1.233194  
1-100210-A-36.wav -0.436426 -1.259930 -1.493075  
1-100210-B-36.wav -0.952827 -1.399003 -1.195466  
1-101296-A-19.wav -0.257224 -1.720606 -3.032294  

[5 rows x 50 columns]

Let's download the predictions file and submit it to the Challenge in Moodle.

## Train Cross Validation

In [53]:
!python train_crossval.py

*****
train folds are {2, 3, 4, 5} and test fold is {1}
random wave cropping

ResNet(block=ResidualBlock, layers=[2, 2, 2, 2], num_classes=config.n_classes)
*****

  0% 0/40 [00:00<?, ?bat/s]/usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()
100% 40/40 [00:36<00:00,  1.11bat/s]
100% 10/10 [00:10<00:00,  1.09s/bat]
  0%|           | 1/200 [00:46<2:35:35, 46.91s/ep] TrnAcc=0.095 ValAcc=0.144 TrnLoss=3.694 ValLoss=3.347 < inf --> checkpoint
 60% 24/40 [00:20<00:13,  1.16bat/s]
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 631, in __next__
    data = self._next_data()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py", line 1329, in _next_data
    idx, data = self._get_data()
  File "/usr/local/lib/python3.10/